pgwire: decouple jwt authentication and authorization logic

shriramters opened the pull request:
Previously, the logic for JWT authorization (extracting group claims and synchronizing roles) was located within the `Authenticator` behavior. This was a necessary design when the JWT token was only exchanged from the `AuthConn` within the authenticator.

This was inadequate because it bundled authentication and authorization logic, violating the separation of concerns intended by the `AuthBehaviors` framework and creating an inconsistency with other methods like `AuthLDAP`. A recent change (#149415) made the token available via closure capture to all behaviors, removing the original constraint and making this refactoring possible.

To address this, this patch decouples the logic. The `Authenticator` is now solely responsible for validating the token (authentication). The authorization logic has been moved to a new `Authorizer` behavior, which aligns the implementation with the framework's design and improves code clarity and maintainability.

Fixes: #150720

Release note: None

shriramters committed:
pgwire: decouple jwt authentication and authorization logic

Previously, the logic for JWT authorization (extracting group claims
and synchronizing roles) was located within the `Authenticator`
behavior. This was a necessary design when the JWT token was only
exchanged from the `AuthConn` within the authenticator.

This was inadequate because it bundled authentication and
authorization logic, violating the separation of concerns intended by
the `AuthBehaviors` framework and creating an inconsistency with other
methods like `AuthLDAP`. A recent change (#149415) made the token
available via closure capture to all behaviors, removing the original
constraint and making this refactoring possible.

To address this, this patch decouples the logic. The `Authenticator`
is now solely responsible for validating the token
(authentication). The authorization logic has been moved to a new
`Authorizer` behavior, which aligns the implementation with the
framework's design and improves code clarity and maintainability.

Fixes: #150720

Release note: None

souravcrl reviewed (commented):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150721)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shriramters)

___
*[`pkg/sql/pgwire/auth_methods.go` line 863 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-OW-WPFZ-jEBPH_gdDjo:-OW-WPFZ-jEBPH_gdDjp:b6b0s8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/67571e0c550112c0ed0cf459710e9b650734ed54/pkg/sql/pgwire/auth_methods.go#L863)):*
> ```Go
> 
> 	b.SetProvisioner(func(ctx context.Context) error {
> 		if validationErr != nil {
> ```

To ensure consistency with authLDAP can we have a log  indicating the start of provisioning?
___
*[`pkg/sql/pgwire/auth_methods.go` line 907 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-OW-W4by72SZ8MPe1WQm:-OW-W4by72SZ8MPe1WQn:bw1bow7) ([raw file](https://github.com/cockroachdb/cockroach/blob/67571e0c550112c0ed0cf459710e9b650734ed54/pkg/sql/pgwire/auth_methods.go#L907)):*
> ```Go
> 			return err
> 		}
> ```

To ensure consistency with authLDAP can we have a log  indicating the start of authentication?
___
*[`pkg/sql/pgwire/auth_methods.go` line 923 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-OW-WIO0C7-Hv65Nj8E_:-OW-WIO0C7-Hv65Nj8Ea:b-ogr316) ([raw file](https://github.com/cockroachdb/cockroach/blob/67571e0c550112c0ed0cf459710e9b650734ed54/pkg/sql/pgwire/auth_methods.go#L923)):*
> ```Go
> 
> 	b.SetAuthorizer(func(ctx context.Context, systemIdentity string, clientConnection bool) error {
> 		// Ask the CCL verifier for groups (nil slice means feature disabled).
> ```

To ensure consistency with authLDAP can we have a log  indicating the start of authorization?


<!-- Sent from Reviewable.io -->

shriramters committed:
pgwire,provisioning: align JWT auth logging and telemetry with LDAP

Previously, the JWT authentication flow lacked specific telemetry
counters for its provisioning process and had less verbose logging
compared to the LDAP authentication flow.

This was inadequate because it created an observability gap, making it
harder to monitor and debug JWT-specific authentication and
provisioning steps compared to the more mature LDAP implementation.

To address this, this patch introduces JWT-specific telemetry counters
for provisioning (`begin`, `success`, `enable`) and adds more detailed
logging messages at each stage of the `authJwtToken` flow
(provisioning, authentication, and authorization). The new logs and
metrics are modeled directly after the existing implementation for
LDAP, ensuring a consistent operational experience for both enterprise
authentication methods.

Release note: None

shriramters reviewed (commented):
@souravcrl I have added Telemetry counters for parity with LDAP in a new commit. 
Previously I was unsure if it is okay to add Telemetry when settings are not reportable. I then saw your comment on the OIDC User Provisioning PR so I added it here as well.
I can remove it if you want that to be added in a separate PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-:-OW-eQJ-9F_qnUfGvc51:bdma8og)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

___
*[`pkg/sql/pgwire/auth_methods.go` line 863 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-OW-WPFZ-jEBPH_gdDjo:-OW-dUBhEfHF4DhJFL8K:bzg48nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/67571e0c550112c0ed0cf459710e9b650734ed54/pkg/sql/pgwire/auth_methods.go#L863)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

To ensure consistency with authLDAP can we have a log  indicating the start of provisioning?
</blockquote></details>

Done, Added. There's an info log down below already so I did it this way.

https://github.com/shriramters/cockroach/blob/bd5d634a9bd643829024a1ebcd87f57b3b3cb7b0/pkg/sql/pgwire/auth_methods.go#L886
___
*[`pkg/sql/pgwire/auth_methods.go` line 907 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-OW-W4by72SZ8MPe1WQm:-OW-eBlc9O30HVlvxnnA:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/67571e0c550112c0ed0cf459710e9b650734ed54/pkg/sql/pgwire/auth_methods.go#L907)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

To ensure consistency with authLDAP can we have a log  indicating the start of authentication?
</blockquote></details>

Done
___
*[`pkg/sql/pgwire/auth_methods.go` line 923 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-OW-WIO0C7-Hv65Nj8E_:-OW-eM5aBmoMWHKxFgO7:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/67571e0c550112c0ed0cf459710e9b650734ed54/pkg/sql/pgwire/auth_methods.go#L923)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

To ensure consistency with authLDAP can we have a log  indicating the start of authorization?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

souravcrl reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150721#-:-OWEgjvv4uzxROe1kMSF:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @shriramters)



<!-- Sent from Reviewable.io -->

shriramters commented:
bors r=souravcrl
